File: packages/validator-php/src/Patterns/Nfa.php
<?php

declare(strict_types=1);

namespace CRUDUI\Validator\Patterns;

use CRUDUI\Validator\Values\CodePointSet;
use CRUDUI\Validator\Values\Utf8;

/**
 * A Thompson NFA that matches whole texts. It is simulated state set by state set:
 * time is proportional to the text length times the number of states, and memory to
 * the number of states.
 *
 * @internal
 */
final class Nfa
{
    /** Consumes one code point of `set` and continues at `next`. */
    public const CHAR = 0;

    /** Continues at both `next` and `alternative`. */
    public const SPLIT = 1;

    /** Continues at `next` without consuming. */
    public const EPSILON = 2;

    /** Accepts. */
    public const MATCH = 3;

    /**
     * @param list<int> $kinds state kinds
     * @param list<int> $setIndexes index into $sets of each CHAR state, else -1
     * @param list<int> $next first successor of each state, else -1
     * @param list<int> $alternatives second successor of each SPLIT state, else -1
     * @param list<CodePointSet> $sets the distinct sets of the pattern
     * @param int $start the start state
     */
    public function __construct(
        private readonly array $kinds,
        private readonly array $setIndexes,
        private readonly array $next,
        private readonly array $alternatives,
        private readonly array $sets,
        private readonly int $start,
    ) {
    }

    /** The number of states. */
    public function stateCount(): int
    {
        return \count($this->kinds);
    }

    /**
     * Whether the whole UTF-8 text is accepted.
     *
     * @throws \InvalidArgumentException when the text is not valid UTF-8
     */
    public function matches(string $text): bool
    {
        $marks = array_fill(0, \count($this->kinds), -1);
        $generation = 0;
        $states = $this->closure([$this->start], $marks, $generation);
        $length = \strlen($text);
        $offset = 0;
        while ($offset < $length) {
            $codePoint = Utf8::next($text, $offset);
            $generation++;
            $members = [];
            $targets = [];
            foreach ($states as $state) {
                if ($this->kinds[$state] !== self::CHAR) {
                    continue;
                }
                $index = $this->setIndexes[$state];
                $member = $members[$index] ??= $this->sets[$index]->contains($codePoint);
                if ($member) {
                    $targets[] = $this->next[$state];
                }
            }
            if ($targets === []) {
                return false;
            }
            $states = $this->closure($targets, $marks, $generation);
        }
        foreach ($states as $state) {
            if ($this->kinds[$state] === self::MATCH) {
                return true;
            }
        }
        return false;
    }

    /**
     * The CHAR and MATCH states reachable from $roots without consuming input.
     *
     * @param list<int> $roots
     * @param list<int> $marks generation of the last visit of each state
     * @return list<int>
     */
    private function closure(array $roots, array &$marks, int $generation): array
    {
        $states = [];
        $stack = array_reverse($roots);
        while ($stack !== []) {
            $state = array_pop($stack);
            if ($marks[$state] === $generation) {
                continue;
            }
            $marks[$state] = $generation;
            switch ($this->kinds[$state]) {
                case self::SPLIT:
                    $stack[] = $this->alternatives[$state];
                    $stack[] = $this->next[$state];
                    break;
                case self::EPSILON:
                    $stack[] = $this->next[$state];
                    break;
                default:
                    $states[] = $state;
            }
        }
        return $states;
    }
}

File: packages/validator-rust/src/validate/pattern/nfa.rs
//! Thompson NFA for a pattern tree, and the whole-text matcher.
//!
//! Matching simulates every state at once: time is proportional to the text
//! length times the number of states, memory to the number of states. Nothing
//! recurses on the text; the epsilon closure uses an explicit stack.

use super::sets::CodeSet;
use super::tree::Node;

/// An NFA state; `usize` fields are state indexes.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
enum State {
    /// Consume a code point in the set, then continue.
    Char {
        set: usize,
        next: usize,
    },
    Split(usize, usize),
    Epsilon(usize),
    Match,
}

/// A compiled pattern.
#[derive(Debug)]
pub(crate) struct Program {
    states: Vec<State>,
    start: usize,
    sets: Vec<CodeSet>,
}

impl Program {
    /// Compile a tree whose atoms index `sets`.
    pub(crate) fn new(root: &Node, sets: Vec<CodeSet>) -> Self {
        let mut compiler = Compiler { states: Vec::new() };
        let accept = compiler.push(State::Match);
        let start = compiler.compile(root, accept);
        Program {
            states: compiler.states,
            start,
            sets,
        }
    }

    /// The number of states.
    #[cfg(test)]
    pub(crate) fn len(&self) -> usize {
        self.states.len()
    }

    /// Whether the whole of `text` matches.
    pub(crate) fn is_match(&self, text: &str) -> bool {
        let mut run = Run {
            program: self,
            marks: vec![0; self.states.len()],
            generation: 1,
            stack: Vec::new(),
        };
        let mut current = Vec::new();
        run.close(self.start, &mut current);
        let mut next = Vec::new();
        // Per-set membership for the current code point: (generation, member).
        let mut membership: Vec<(u64, bool)> = vec![(0, false); self.sets.len()];
        for c in text.chars() {
            run.generation += 1;
            let generation = run.generation;
            next.clear();
            for &state in &current {
                if let State::Char { set, next: to } = self.states[state] {
                    let (seen, member) = &mut membership[set];
                    if *seen != generation {
                        *seen = generation;
                        *member = self.sets[set].contains(c);
                    }
                    if *member {
                        run.close(to, &mut next);
                    }
                }
            }
            std::mem::swap(&mut current, &mut next);
            if current.is_empty() {
                return false;
            }
        }
        current
            .iter()
            .any(|&state| self.states[state] == State::Match)
    }
}

/// The state of one match.
struct Run<'a> {
    program: &'a Program,
    /// The generation in which each state was last added.
    marks: Vec<u64>,
    generation: u64,
    stack: Vec<usize>,
}

impl Run<'_> {
    /// Add the epsilon closure of `state` to `list`, keeping only states that
    /// consume a code point or accept.
    fn close(&mut self, state: usize, list: &mut Vec<usize>) {
        self.stack.push(state);
        while let Some(state) = self.stack.pop() {
            if self.marks[state] == self.generation {
                continue;
            }
            self.marks[state] = self.generation;
            match self.program.states[state] {
                State::Char { .. } | State::Match => list.push(state),
                State::Epsilon(next) => self.stack.push(next),
                State::Split(first, second) => {
                    self.stack.push(second);
                    self.stack.push(first);
                }
            }
        }
    }
}

struct Compiler {
    states: Vec<State>,
}

impl Compiler {
    fn push(&mut self, state: State) -> usize {
        self.states.push(state);
        self.states.len() - 1
    }

    /// Compile `node` so that it continues to `next`; returns its entry state.
    /// A node of size 0 can only match empty text and becomes one epsilon.
    fn compile(&mut self, node: &Node, next: usize) -> usize {
        if node.size() == 0 {
            return self.push(State::Epsilon(next));
        }
        match node {
            Node::Atom(set) => self.push(State::Char { set: *set, next }),
            Node::Sequence(parts) => parts
                .iter()
                .rev()
                .filter(|part| part.size() > 0)
                .fold(next, |entry, part| self.compile(part, entry)),
            Node::Alternation(alternatives) => {
                let mut entries: Vec<usize> = alternatives
                    .iter()
                    .filter(|alternative| alternative.size() > 0)
                    .map(|alternative| self.compile(alternative, next))
                    .collect();
                if entries.len() < alternatives.len() {
                    // Every empty alternative matches only empty text.
                    entries.push(self.push(State::Epsilon(next)));
                }
                let last = entries.pop().expect("an alternative of positive size");
                entries
                    .into_iter()
                    .rev()
                    .fold(last, |rest, entry| self.push(State::Split(entry, rest)))
            }
            Node::Repeat { item, min, max } => {
                let mut entry = match max {
                    None => {
                        // One loop: the item repeats any number of times.
                        let split = self.push(State::Split(0, next));
                        let body = self.compile(item, split);
                        self.states[split] = State::Split(body, next);
                        split
                    }
                    Some(max) => {
                        // (max − min) nested optional copies.
                        let mut entry = next;
                        for _ in *min..*max {
                            let body = self.compile(item, entry);
                            entry = self.push(State::Split(body, next));
                        }
                        entry
                    }
                };
                for _ in 0..*min {
                    entry = self.compile(item, entry);
                }
                entry
            }
        }
    }
}
